### ISSUE TITLE
Features not matching version after an upgrade to 8.13+

### ISSUE THREAD

**Issue by thbkrkr** [elastic/elasticsearch#109254]
### Elasticsearch Version

8.13.x

- Java Version: _bundled_

- OS Version: N/A (different k8s versions)

### Problem Description

ECK operator `2.12.1`  fails upgrading Elasticsearch `8.13+` as it is stalled on the following error when calling the desired nodes API:
```rust
400 Bad Request: {
    Reason: [node_version] field is required and must have a valid value 
    Type: x_content_parse_exception
    ...
}
```

### Steps to Reproduce

The steps should be (not tested yet):
- Deploy ECK 2.12.1+
- Deploy Elasticsearch 8.11.x?
- Upgrade to Elasticsearch 8.13+

### Notes

Note: this is from ECK `2.12.1` that the operator stops to use the deprecated `node_version` field if the cluster is 8.13+:
- https://github.com/elastic/cloud-on-k8s/pull/7663

Two occurences of this issue have been reported for these versions upgrades:
- 8.11.x -> 8.13.3
- 8.11.4 -> 8.13.2

Each time, the users confirmed that:
- some nodes did not have `desired_node.version_deprecated` feature 
- a restart fixes the issue

**Comment by thecoop** [comment]
@thbkrkr To help debug this, it'll be really helpful to have the order that nodes were upgraded, the next time this occurs, as well as logs from all the upgraded nodes.

**Comment by thecoop** [comment]
We already have some accounting to make sure features are set properly when a master node is upgraded, but it looks like this is going awry

**Comment by thecoop** [comment]
This does not reproduce readily, I suspect it very much depends on the order nodes are restarted. Continuing to investigate.

**Comment by thecoop** [comment]
I just need all the logs of the nodes from when the upgrade started to when it completed

**Comment by thecoop** [comment]
@mikeprince3 Just spotted you're not an Elastician, so don't have access to our infrastructure. If you can provide a tarball of the log files somewhere that would be very helpful; if you don't have anywhere to upload them to please let me know and we can sort something out.

**Comment by thecoop** [comment]
Just the ES nodes will do - this is a bug with elasticsearch

**Comment by thecoop** [comment]
@mikeprince3 Thanks for the logs. This bug is around the exact order in which nodes are restarted and get elected to master, so could you send the logs for the other two master nodes, and maybe one node that was unaffected by this bug for comparison?

**Comment by thecoop** [comment]
Looks like this bug happens when there are non-master-eligible nodes in a cluster when a cluster is upgraded. The non-master-eligible nodes are not going through the same codepath in `NodeJoinExecutor` the upgraded master as the master-eligible nodes, leading to features not being upgraded as expected. Restarting the relevant nodes afterwards lead to the correct features getting upgraded.

**Comment by thecoop** [comment]
Thanks for the logs @mikeprince3, we've merged in a fix that will be in 8.15
